Set the location for the candidate set when looking up C++ operators

bricknerb opened the pull request:
This adds location information and prevents crashes in some cases of template instantiation in operator lookup.

Removed `InCppOperatorLookup` note as it is no longer necessary.

Part of #5995.

bricknerb committed:
Set the location for the candidate set when looking up C++ operators
Couldn't generate a test that covers this.

zygoloid reviewed (commented):
I think you should be able to test this by triggering a diagnostic from a template instantiation performed for the lookup. Eg:
```
import Cpp inline '''
template<typename T> struct B {
};
template<typename T> struct A {
  friend typename T::error operator+(B<A>, B<A>);
};
using X = B<A<int>>;
''';
fn F(x: Cpp.X) {
  x + x;
}
```

zygoloid commented:
For the example I gave, we [currently produce](https://godbolt.org/z/h4oa9WTdr):
```console
<source>:5:19: error: type 'int' cannot be used prior to '::' because it has no members
    5 |   friend typename T::error operator+(B<A>, B<A>);
      |                   ^
note: in instantiation of template class 'A<int>' requested here
<source>:10:3: note: in `Cpp` operator `AddWith` lookup
  x + x;
  ^~~~~
```
Two things of note here:
1) The `note: in instantiation of` diagnostic has no location. I believe this will have a proper location after this change.
2) The ``note: in `Cpp` operator `AddWith` lookup`` note should be unnecessary after this change, as any diagnostic produced by Clang should already have the location attached, so we can probably stop adding that note. We still need to create the `AnnotationScope`, however, because it flushes the Clang diagnostics to the Carbon diagnostic stream in its destructor. See [the corresponding code in overload resolution](https://github.com/carbon-language/carbon-lang/blob/83ba714165c6989d69e6ecff0e1965a267af7e66/toolchain/check/cpp/overload_resolution.cpp#L98), which has a comment you can copy here.

bricknerb committed:
Merge branch 'main' into operator7

bricknerb committed:
Add template instantiation tests that demonstrate setting the location properly fixes a crash in some cases.

bricknerb committed:
Remove the no longer necessary `InCppOperatorLookup` note.

bricknerb commented:
> For the example I gave, we [currently produce](https://godbolt.org/z/h4oa9WTdr):
> 
> ```
> <source>:5:19: error: type 'int' cannot be used prior to '::' because it has no members
>     5 |   friend typename T::error operator+(B<A>, B<A>);
>       |                   ^
> note: in instantiation of template class 'A<int>' requested here
> <source>:10:3: note: in `Cpp` operator `AddWith` lookup
>   x + x;
>   ^~~~~
> ```
> 

Thanks!
This is very helpful.
Trying to work with this example, I've noticed that while it seems to work in Compiler Explorer, it crashes when I run this locally (without this change).
I've simplified it to make it a fully linkable example that crashes for me:

```
// main.carbon

library "Main";

import Cpp inline '''
template<typename T> struct B {};
template<typename T> struct A {};
using X = B<A<int>>;
int operator+(X, X) { return 0; }
''';
fn F(x: Cpp.X) -> i32 {
  return x + x;
}

fn Run() -> i32 {
  return F(Cpp.X.B());
}
```

This crashes for me while it works in Compiler Explorer: https://godbolt.org/z/3Y8WGqh49.
I couldn't figure why it crashes for me but not in Compiler Explorer.

Given that this change fixes this crash, I've added this as a test as well.
However, I couldn't check the impact on the compiler error, just that it fixes the crash.

> Two things of note here:
> 
> 1. The `note: in instantiation of` diagnostic has no location. I believe this will have a proper location after this change.

I do see the `note` has a location with this change.

> 2. The `` note: in `Cpp` operator `AddWith` lookup `` note should be unnecessary after this change, as any diagnostic produced by Clang should already have the location attached, so we can probably stop adding that note. We still need to create the `AnnotationScope`, however, because it flushes the Clang diagnostics to the Carbon diagnostic stream in its destructor. See [the corresponding code in overload resolution](https://github.com/carbon-language/carbon-lang/blob/83ba714165c6989d69e6ecff0e1965a267af7e66/toolchain/check/cpp/overload_resolution.cpp#L98), which has a comment you can copy here.

Done.

zygoloid reviewed (approved):
Thanks!
